### PR TITLE
ImageNet Example

### PR DESCRIPTION
**Pull request by zphang**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [X] Did you read the [contributor guideline](https://github.com/williamFalcon/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Addresses #475

----

This PR is more or less a direct port of https://github.com/pytorch/examples/tree/master/imagenet to Lightning. I tried to change the logic from the original as little as possible so a reader can compare both easily.

Still new to the framework, so pardon me for any obvious mistakes or incorrect of features. E.g. Based on `gpu_example.py`, I'm not entirely clear what hparams should belong to the LightningModule and which to be in the parent parser.

### CODE REVIEW

**Comment by zphang** [comment]
Addressed the comments, except for the DistributedSampler one.

**Comment by neggert** [comment]
Thanks. Looks good to me now, aside from that question about ddp2.

File: pl_examples/full_examples/imagenet/imagenet_example.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+            'log': tqdm_dict
+        })
+
+        return output
+
+    def validation_step(self, batch, batch_idx):
+        images, target = batch
+        output = self.model(images)
+        loss_val = F.cross_entropy(output, target)
+        acc1, acc5 = self.__accuracy(output, target, topk=(1, 5))
```
Comment (Borda) [inline pl_examples/full_examples/imagenet/imagenet_example.py]: is this valid? `self.__accuracy()` return list of size `max(topk)` which is 5 here but unpacked variables are just two...

File: pl_examples/full_examples/imagenet/imagenet_example.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            'val_loss': loss_val,
+            'val_acc1': acc1,
+            'val_acc5': acc5,
+        })
+
+        return output
+
+    def validation_end(self, outputs):
+
+        val_loss_mean = 0
```
Comment (Borda) [inline pl_examples/full_examples/imagenet/imagenet_example.py]: the three are very same, you may define it as a local function inside `validation_end`:
```
...
def _compute_val(name):
    val_mean = 0
    for output in outputs:
            val = output[name]
            # reduce manually when using dp
            if self.trainer.use_dp or self.trainer.use_ddp2:
                val = torch.mean(val)
            val_mean += val
    val_mean /= len(outputs)

tqdm_dict = {k: _compute_val(k) for k in ('val_loss', 'val_acc1', 'val_acc5')}
...
```

File: pl_examples/full_examples/imagenet/imagenet_example.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+            lr=self.hparams.lr,
+            momentum=self.hparams.momentum,
+            weight_decay=self.hparams.weight_decay
+        )
+        scheduler = lr_scheduler.ExponentialLR(optimizer, gamma=0.1)
+        return [optimizer], [scheduler]
+
+    @pl.data_loader
+    def train_dataloader(self):
+        train_dir = os.path.join(self.hparams.data, 'train')
```
Comment (Borda) [inline pl_examples/full_examples/imagenet/imagenet_example.py]: is it always "train" sub-folder or shall it be a parameter (with default)?

File: pl_examples/full_examples/imagenet/imagenet_example.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                            metavar='W', help='weight decay (default: 1e-4)',
+                            dest='weight_decay')
+        parser.add_argument('--pretrained', dest='pretrained', action='store_true',
+                            help='use pre-trained model')
+        return parser
+
+
+def get_args():
+    parent_parser = argparse.ArgumentParser(add_help=False)
+    parent_parser.add_argument('data', metavar='DIR',
```
Comment (Borda) [inline pl_examples/full_examples/imagenet/imagenet_example.py]: rather '--data_path' ? also `type=str`

File: pl_examples/full_examples/imagenet/imagenet_example.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                               help='path to dataset')
+    parent_parser.add_argument('--save_path', metavar='DIR', default=".",
+                               help='path to save output')
+    parent_parser.add_argument('--gpus', type=int, default=1,
+                               help='how many gpus')
+    parent_parser.add_argument('--distributed_backend', type=str, default='dp',
```
Comment (Borda) [inline pl_examples/full_examples/imagenet/imagenet_example.py]: consider adding `options=('dp', 'ddp', 'ddp2')`

File: pl_examples/full_examples/imagenet/imagenet_example.py (hunk 1 above)
Comment (zphang) [reply pl_examples/full_examples/imagenet/imagenet_example.py]: It should return the same number of elements as the length of `topk`. Also, the method is taken directly from pytorch_examples.

File: pl_examples/full_examples/imagenet/imagenet_example.py (hunk 3 above)
Comment (zphang) [reply pl_examples/full_examples/imagenet/imagenet_example.py]: I think the folders being named train/val is the standard convention. I don't feel strongly either way, but I think having three arguments for data_path/train_data_path (optional)/val_data_path (optional) might be excessive.

File: pl_examples/full_examples/imagenet/imagenet_example.py (hunk 1 above)
Comment (Borda) [reply pl_examples/full_examples/imagenet/imagenet_example.py]:
```
topk = (1, 5)
max(topk)
Out[2]: 5
len(topk)
Out[3]: 2
```

File: pl_examples/full_examples/imagenet/imagenet_example.py (hunk 2 above)
Comment (zphang) [reply pl_examples/full_examples/imagenet/imagenet_example.py]: Converted to a loop instead.

**Comment by zphang** [comment]
Addressed more feedback!

**Comment by williamFalcon** [comment]
@zphang sweet. Great addition.

File: pl_examples/full_examples/imagenet/imagenet_example.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            return res
+
+    def configure_optimizers(self):
+        optimizer = optim.SGD(
+            self.parameters(),
+            lr=self.hparams.lr,
+            momentum=self.hparams.momentum,
+            weight_decay=self.hparams.weight_decay
+        )
+        scheduler = lr_scheduler.ExponentialLR(optimizer, gamma=0.1)
```
Comment (SohamTamba) [inline pl_examples/full_examples/imagenet/imagenet_example.py:131]: Hi,
I think this should be StepLR with period=30 (or total_epochs//3)

**Pull request merged**